services/slack_grid/parse.go

```diff
@@ -0,0 +1,383 @@
+package slack_grid
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v6/model"
+	slack "github.com/mattermost/mmetl/services/slack"
+	"github.com/pkg/errors"
+
+	log "github.com/sirupsen/logrus"
+)
+
+type ChannelFiles string
+
+const (
+	ChannelFilePublic  ChannelFiles = "channels"
+	ChannelFilePrivate ChannelFiles = "groups"
+	ChannelFileDM      ChannelFiles = "dms"
+	ChannelFileGM      ChannelFiles = "mpims"
+)
+
+const ErrFindingTeamID = "Could not find team ID for channel %v, ID: %v"
+const ErrFindingTeamName = "Could not find team name for channel %v"
+
+type BulkSlackExport struct {
+	Public  []slack.SlackChannel
+	Private []slack.SlackChannel
+	GMs     []slack.SlackChannel
+	DMs     []slack.SlackChannel
+}
+
+type Post struct {
+	*slack.SlackPost
+	Team       string `json:"team"`
+	UserTeam   string `json:"user_team"`
+	SourceTeam string `json:"source_team"`
+}
+
+type Channel struct {
+	*slack.SlackChannel
+	Team string `json:"team"`
+}
+
+type BulkTransformer struct {
+	slack.Transformer
+	Teams map[string]string
+
+	// the root path to export slack to for parsing.
+	// defaults to ./tmp/slack_grid
+	dirPath string
+	pwd     string
+}
+
+type ChannelsToMove struct {
+	SlackChannel slack.SlackChannel
+	TeamID       string
+	TeamName     string
+	Moved        bool
+
+	// Path is the name of the channel or the channel ID for DMs
+	Path string
+}
+
+func NewGridTransformer(logger log.FieldLogger) *BulkTransformer {
+	return &BulkTransformer{
+		Transformer: slack.Transformer{
+			Intermediate: &slack.Intermediate{},
+			Logger:       logger,
+		},
+	}
+}
+
+func (t *BulkTransformer) ParseBulkSlackExportFile(zipReader *zip.Reader) (*BulkSlackExport, error) {
+	slackExport := BulkSlackExport{}
+
+	// only finding the root information here and storing it.
+	for i, file := range zipReader.File {
+		if file.FileInfo().IsDir() || strings.Contains(file.Name, "/") {
+			continue
+		}
+		err := func(i int, file *zip.File) error {
+			reader, err := file.Open()
+			if err != nil {
+				return errors.Wrap(err, "Error opening file for parsing.")
+			}
+			defer reader.Close()
+
+			switch file.Name {
+			case string(ChannelFilePublic) + ".json":
```

Comment:
Probably not a huge deal, but we are unnecessarily creating new strings every time this function runs. Maybe we can precompute these right next to the definition of `ChannelFilePublic`?